File: problems/cpp/paint-house.cc
// 3 ms, 09/15/2017
// Time  : O(n * k), k = 3
// Space : O(n * k), k = 3
class Solution {
public:
    int minCost(vector<vector<int>>& costs)
    {
        int n = costs.size();
        if (n == 0)
            return 0;
        for (int i = 1; i < n; i++) {
            costs[i][0] += min(costs[i - 1][1],costs[i - 1][2]);
            costs[i][1] += min(costs[i - 1][0],costs[i - 1][2]);
            costs[i][2] += min(costs[i - 1][0],costs[i - 1][1]);
        }
        return *min_element(costs[n-1].begin(), costs[n-1].end());
    }
};

File: code/python/paint-house.py
# Date  : 03/17/2018
# Time  : O(m*n), m = number of colors (3), n = number of houses
# Space : O(m*n)

class Solution:
    def minCost(self, costs):
        """
        :type costs: List[List[int]]
        :rtype: int
        """
        if not costs:
            return 0

        for h in range(1, len(costs)):
            costs[h][0] += min(costs[h-1][1], costs[h-1][2])
            costs[h][1] += min(costs[h-1][0], costs[h-1][2])
            costs[h][2] += min(costs[h-1][0], costs[h-1][1])

        return min(costs[-1])
